src/indexes.rs

```diff
@@ -80,6 +80,25 @@ where
         };
         (l, Some(l))
     }
+
+    fn fold<B, F>(self, init: B, mut f: F) -> B
+    where
+        F: FnMut(B, D::Pattern) -> B,
+    {
+        let IndicesIter { mut index, dim } = self;
+        let inner_axis = dim.ndim() - 1;
```

Comment:
We need to handle ndim zero somehow

Reply:
Good catch. I should have tested this more thoroughly.

Reply:
I also found a bug for `ndim > 2`: `.next_for()` requires requires the input index to be in-bounds, but my original implementation didn't keep the index in-bounds. I've added more tests, and I think everything should be fixed now. I'm still getting the same performance improvement in the fixed version.

Reply:
No worries. It's the tiniest things. And shows how much tests we need to cover it all..